hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AdjustFungibleTokenChangesStep.java

```diff
@@ -0,0 +1,219 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.token.impl.handlers.transfer;
+
+import static com.hedera.hapi.node.base.ResponseCodeEnum.ACCOUNT_AMOUNT_TRANSFERS_ONLY_ALLOWED_FOR_FUNGIBLE_COMMON;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.AMOUNT_EXCEEDS_ALLOWANCE;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.INVALID_ACCOUNT_ID;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.SPENDER_DOES_NOT_HAVE_ALLOWANCE;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.UNEXPECTED_TOKEN_DECIMALS;
+import static com.hedera.node.app.service.token.impl.handlers.BaseCryptoHandler.asAccount;
+import static com.hedera.node.app.service.token.impl.util.TokenHandlerHelper.getIfUsable;
+import static com.hedera.node.app.spi.workflows.HandleException.validateTrue;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.base.TokenType;
+import com.hedera.hapi.node.token.CryptoTransferTransactionBody;
+import com.hedera.node.app.service.mono.utils.EntityNumPair;
+import com.hedera.node.app.service.token.impl.WritableAccountStore;
+import com.hedera.node.app.service.token.impl.WritableTokenRelationStore;
+import com.hedera.node.app.service.token.impl.WritableTokenStore;
+import com.hedera.node.app.service.token.impl.handlers.BaseTokenHandler;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Puts all fungible token changes from CryptoTransfer into state's modifications map.
+ */
+public class AdjustFungibleTokenChangesStep extends BaseTokenHandler implements TransferStep {
+    // The CryptoTransferTransactionBody here is obtained by replacing aliases with their
+    // corresponding accountIds.
+    private final CryptoTransferTransactionBody op;
+    private final AccountID topLevelPayer;
+
+    public AdjustFungibleTokenChangesStep(
+            @NonNull final CryptoTransferTransactionBody op, @NonNull final AccountID topLevelPayer) {
+        requireNonNull(op);
+        requireNonNull(topLevelPayer);
+
+        this.op = op;
+        this.topLevelPayer = topLevelPayer;
+    }
+
+    @Override
+    public void doIn(@NonNull final TransferContext transferContext) {
+        requireNonNull(transferContext);
+
+        final var handleContext = transferContext.getHandleContext();
+        final var tokenStore = handleContext.writableStore(WritableTokenStore.class);
+        final var tokenRelStore = handleContext.writableStore(WritableTokenRelationStore.class);
+        final var accountStore = handleContext.writableStore(WritableAccountStore.class);
+
+        // two maps for aggregating the changes to the token balances and allowances.
```

Comment:
Remind me why we need to do aggregation at this stage?

Is it just because there can be both an `is_approval=true` and `is_approval=false` version of a `(0.0.<ACCOUNT>, 0.0.<TOKEN>)` adjustment?

Reply:
yeah exactly. For either one balance should be changed, but for `is_approval=true` allowances should be modified